habitat_baselines/evaluate_ppo.py

```diff
@@ -149,13 +219,82 @@ def main():
         episode_counts += 1 - not_done_masks
         current_episode_reward *= not_done_masks
 
+        # added for video support
+        next_episodes = envs.current_episodes()
+        envs_to_pause = []
+        n_envs = envs.num_envs
+        for i in range(n_envs):
+            if next_episodes[i].episode_id in stats_episodes:
+                envs_to_pause.append(i)
+
+            if not_done_masks[i].item() == 0:
+                # episode ended, record and generate videos as specified
+                stats_episodes.add(current_episodes[i].episode_id)
+                if args.video_option != 0 and len(rgb_frames[i]) > 0:
+                    video_name = "episode{}_ckpt{}_spl{:.2f}".format(
+                        current_episodes[i].episode_id,
+                        checkpoint_idx,
+                        infos[i]["spl"],
+                    )
+                    if args.video_option in [1, 3]:
+                        images_to_video(
+                            rgb_frames[i], args.video_dir, video_name
+                        )
+                    if args.video_option in [2, 3]:
+                        frames_to_tb_video(
+                            "episode{}".format(current_episodes[i].episode_id),
+                            checkpoint_idx,
+                            rgb_frames[i],
+                            writer,
+                            fps=10,
+                        )
+                    rgb_frames[i] = []
+
+            elif args.video_option != 0:
+                # episode continues, record current frame
+                frame = generate_frame(observations[i], infos[i])
+                rgb_frames[i].append(frame)
+
+        if len(envs_to_pause) > 0:
+            # stop tracking ended episodes
+
+            state_index = list(range(envs.num_envs))
+            for idx in reversed(envs_to_pause):
+                state_index.pop(idx)
+                envs.pause_at(idx)
+
+            # indexing along the batch dimensions
+            test_recurrent_hidden_states = test_recurrent_hidden_states[
```

Comment:
Be sure to double check this properly in your local test, it is very easy to make a mistake here, the code looks correct.